File: protocol/elixir/test/dust_protocol/glob_test.exs
defmodule DustProtocol.GlobTest do
  use ExUnit.Case, async: true

  alias DustProtocol.Glob

  describe "compile/1" do
    test "compiles a segment-list pattern" do
      assert {:ok, {:compiled, _}} = Glob.compile(["posts", "*"])
    end

    test "compiles a rendered-string pattern" do
      assert {:ok, {:compiled, _}} = Glob.compile("posts/*")
    end

    test "rejects empty pattern" do
      assert Glob.compile([]) == {:error, :empty_path}
      assert Glob.compile("") == {:error, :empty_path}
    end

    test "rejects empty segment" do
      assert Glob.compile(["a", ""]) == {:error, :empty_segment}
      assert Glob.compile("a//b") == {:error, :empty_segment}
    end

    test "rejects ** in non-tail position" do
      assert Glob.compile(["a", "**", "b"]) == {:error, :wildcard_many_not_tail}
      assert Glob.compile("a/**/b") == {:error, :wildcard_many_not_tail}
    end

    test "accepts ** in tail position" do
      assert {:ok, _} = Glob.compile(["a", "**"])
      assert {:ok, _} = Glob.compile("a/**")
    end

    test "rejects invalid path escapes" do
      assert Glob.compile("a~") == {:error, :invalid_escape}
    end
  end

  describe "match?/2 — wildcards" do
    test "* matches exactly one segment" do
      assert Glob.match?(["posts", "*"], ["posts", "hello"])
      refute Glob.match?(["posts", "*"], ["posts"])
      refute Glob.match?(["posts", "*"], ["posts", "hello", "title"])
    end

    test "** at tail matches one-or-more segments" do
      assert Glob.match?(["posts", "**"], ["posts", "hello"])
      assert Glob.match?(["posts", "**"], ["posts", "a", "b", "c"])
      refute Glob.match?(["posts", "**"], ["posts"])
    end

    test "literal segments must match exactly" do
      assert Glob.match?(["a", "b", "c"], ["a", "b", "c"])
      refute Glob.match?(["a", "b", "c"], ["a", "b", "d"])
    end

    test "* mid-pattern" do
      assert Glob.match?(["a", "*", "c"], ["a", "b", "c"])
      refute Glob.match?(["a", "*", "c"], ["a", "b", "d"])
    end
  end

  describe "match?/2 — literal characters" do
    test "dots are literal" do
      assert Glob.match?(["hello.world"], ["hello.world"])
      refute Glob.match?(["hello.world"], ["hello", "world"])
    end

    test "slashes in segments work (need ~1 in rendered form)" do
      assert Glob.match?(["files", "image/file"], ["files", "image/file"])
      assert Glob.match?("files/image~1file", ["files", "image/file"])
    end

    test "tildes in segments work" do
      assert Glob.match?(["a~b"], ["a~b"])
      assert Glob.match?("a~0b", ["a~b"])
    end
  end

  describe "match?/2 — literal wildcard escapes" do
    test "\\* matches literal asterisk segment" do
      assert Glob.match?(["a", "\\*"], ["a", "*"])
      refute Glob.match?(["a", "\\*"], ["a", "x"])
    end

    test "\\** matches literal double-asterisk segment" do
      assert Glob.match?(["a", "\\**"], ["a", "**"])
      refute Glob.match?(["a", "\\**"], ["a", "x", "y"])
    end
  end

  describe "match?/2 — accepts rendered string patterns" do
    test "rendered string is compiled on the fly" do
      assert Glob.match?("posts/*", ["posts", "hello"])
      refute Glob.match?("posts/*", ["posts", "hello", "title"])
    end

    test "compile error in match? raises" do
      assert_raise ArgumentError, ~r/invalid glob pattern/, fn ->
        Glob.match?("a/**/b", ["a", "b"])
      end
    end
  end

  describe "compile!/1" do
    test "returns the compiled pattern" do
      assert {:compiled, _} = Glob.compile!(["a", "*"])
    end

    test "raises on invalid pattern" do
      assert_raise ArgumentError, ~r/invalid glob pattern/, fn ->
        Glob.compile!(["a", "**", "b"])
      end
    end
  end

  describe "fixture conformance (protocol/spec/fixtures/glob_vectors.json)" do
    @fixture_path Elixir.Path.expand(
                    Elixir.Path.join([
                      __DIR__,
                      "..",
                      "..",
                      "..",
                      "spec",
                      "fixtures",
                      "glob_vectors.json"
                    ])
                  )

    @vectors @fixture_path |> File.read!() |> :json.decode()

    for {vector, idx} <- Enum.with_index(@vectors) do
      case vector do
        %{
          "valid" => true,
          "pattern_segments" => p_segs,
          "pattern_rendered" => p_str,
          "path" => path,
          "match" => expected
        } ->
          @vector_idx idx
          @p_segs p_segs
          @p_str p_str
          @path path
          @expected expected

          test "vector ##{@vector_idx}: #{inspect(@p_segs)} vs #{inspect(@path)} -> #{@expected}" do
            # Both forms must give the same result.
            seg_result = Glob.match?(@p_segs, @path)
            str_result = Glob.match?(@p_str, @path)
            assert seg_result == @expected
            assert str_result == @expected
          end

        %{"valid" => false, "pattern_rendered" => p_str, "error" => error} ->
          @vector_idx idx
          @p_str p_str
          @vector_error String.to_atom(error)

          test "vector ##{@vector_idx}: pattern #{inspect(@p_str)} rejected with :#{@vector_error}" do
            assert {:error, @vector_error} = Glob.compile(@p_str)
          end
      end
    end
  end
end

File: sdk/typescript/test/glob.test.ts
import { describe, it, expect } from 'vitest'
import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { compile, match } from '../src/glob'

describe('compile', () => {
  it('compiles a segment array', () => {
    expect(compile(['posts', '*'])).toMatchObject({ tokens: expect.any(Array) })
  })

  it('compiles a rendered slash string', () => {
    expect(compile('posts/*')).toMatchObject({ tokens: expect.any(Array) })
  })

  it('rejects empty pattern', () => {
    expect(() => compile([])).toThrow()
    expect(() => compile('')).toThrow()
  })

  it('rejects ** in non-tail position', () => {
    expect(() => compile(['a', '**', 'b'])).toThrow(/tail position/)
    expect(() => compile('a/**/b')).toThrow(/tail position/)
  })

  it('accepts ** in tail position', () => {
    expect(() => compile(['a', '**'])).not.toThrow()
    expect(() => compile('a/**')).not.toThrow()
  })
})

describe('match — wildcards', () => {
  it('* matches exactly one segment', () => {
    expect(match(['posts', '*'], ['posts', 'hello'])).toBe(true)
    expect(match(['posts', '*'], ['posts'])).toBe(false)
    expect(match(['posts', '*'], ['posts', 'hello', 'title'])).toBe(false)
  })

  it('tail ** matches one-or-more segments', () => {
    expect(match(['posts', '**'], ['posts', 'a'])).toBe(true)
    expect(match(['posts', '**'], ['posts', 'a', 'b', 'c'])).toBe(true)
    expect(match(['posts', '**'], ['posts'])).toBe(false)
  })

  it('literal segments match exactly', () => {
    expect(match(['a', 'b', 'c'], ['a', 'b', 'c'])).toBe(true)
    expect(match(['a', 'b', 'c'], ['a', 'b', 'd'])).toBe(false)
  })

  it('* mid-pattern', () => {
    expect(match(['a', '*', 'c'], ['a', 'b', 'c'])).toBe(true)
    expect(match(['a', '*', 'c'], ['a', 'b', 'd'])).toBe(false)
  })
})

describe('match — literal characters', () => {
  it('dots are literal', () => {
    expect(match(['hello.world'], ['hello.world'])).toBe(true)
    expect(match(['hello.world'], ['hello', 'world'])).toBe(false)
  })

  it('rendered string with escapes round-trips', () => {
    expect(match('files/image~1file', ['files', 'image/file'])).toBe(true)
    expect(match('a~0b', ['a~b'])).toBe(true)
  })
})

describe('match — literal-wildcard escapes', () => {
  it('\\* matches literal asterisk segment', () => {
    expect(match(['a', '\\*'], ['a', '*'])).toBe(true)
    expect(match(['a', '\\*'], ['a', 'x'])).toBe(false)
  })

  it('\\** matches literal double-asterisk segment', () => {
    expect(match(['a', '\\**'], ['a', '**'])).toBe(true)
    expect(match(['a', '\\**'], ['a', 'x', 'y'])).toBe(false)
  })
})

// ----------------------------------------------------------------------
// Fixture-driven conformance against the canonical protocol package.
// ----------------------------------------------------------------------

interface MatchVector {
  valid: true
  pattern_segments: string[]
  pattern_rendered: string
  path: string[]
  match: boolean
  note?: string
}

interface InvalidPatternVector {
  valid: false
  pattern_rendered: string
  error: string
}

type GlobVector = MatchVector | InvalidPatternVector

const fixturePath = resolve(__dirname, '../../../protocol/spec/fixtures/glob_vectors.json')
const vectors: GlobVector[] = JSON.parse(readFileSync(fixturePath, 'utf-8'))

describe('fixture conformance (protocol/spec/fixtures/glob_vectors.json)', () => {
  for (const [idx, v] of vectors.entries()) {
    if (v.valid) {
      it(`#${idx}: ${JSON.stringify(v.pattern_segments)} vs ${JSON.stringify(v.path)} → ${v.match}`, () => {
        // Both pattern forms must give the same result.
        expect(match(v.pattern_segments, v.path)).toBe(v.match)
        expect(match(v.pattern_rendered, v.path)).toBe(v.match)
      })
    } else {
      it(`#${idx}: pattern ${JSON.stringify(v.pattern_rendered)} rejected (${v.error})`, () => {
        expect(() => compile(v.pattern_rendered)).toThrow()
      })
    }
  }
})
